### PR TITLE
Add custom channels liquidity edge cases itest

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
This PR adds an itest that exposes the liquidity related edge cases of custom channels.

The following PRs need to be merged first, in order to bump the [dependencies](https://github.com/lightninglabs/lightning-terminal/pull/842/commits/bf60d5b0d9bc1e80182d80b5882dea48c94ecbba) on this PR:
- [x]  [LND #9092](https://github.com/lightningnetwork/lnd/pull/9092)
- [x]  [Tapd #1120](https://github.com/lightninglabs/taproot-assets/pull/1120)

Closes https://github.com/lightninglabs/taproot-assets/issues/1060
Closes https://github.com/lightninglabs/taproot-assets/issues/1073
~Closes https://github.com/lightninglabs/taproot-assets/issues/1013~

### CODE REVIEW

File: itest/litd_custom_channels_test.go [hunk 1]
```diff
@@ -1675,3 +1678,207 @@ func testCustomChannelsBreach(_ context.Context, net *NetworkHarness,
...
+			fn.None[int64](), lnrpc.Payment_FAILED,
+		)
+
+		done <- true
+	}()
+
+	select {
+	case <-done:
+	case <-timeoutChan:
+		t.Fatalf("Payment didn't fail within expected time duration")
```
Comment (GeorgeTsagk) [inline itest/litd_custom_channels_test.go:1837]: I'm doing this because both timeout and actual payment error result in `lnrpc.Payment_FAILED` above, so I wanna make sure we failed but not because we got stuck in a loop (behavior prior to bug fix)

File: go.mod [hunk 2]
```diff
@@ -222,6 +222,10 @@ require (
 
 replace github.com/lightninglabs/lightning-terminal/autopilotserverrpc => ./autopilotserverrpc
 
+replace github.com/lightningnetwork/lnd => ../lnd
```
Comment (guggero) [inline go.mod:225]: Should update these to your fork's commits so we can see the CI run.

File: itest/assets_test.go [hunk 3]
```diff
@@ -45,6 +46,9 @@ import (
 	"gopkg.in/macaroon.v2"
 )
 
+// PaymentTimeoutSec is the default payment timeout we use in our tests.
+const PaymentTimeoutSec = 6
```
Comment (guggero) [inline itest/assets_test.go:50]: nit: I think we should use a `time.Duration` here, and then juse `Seconds()` further down. Then we don't need the unit in the variable name.

File: itest/litd_custom_channels_test.go [hunk 4]
```diff
@@ -1675,3 +1678,207 @@ func testCustomChannelsBreach(_ context.Context, net *NetworkHarness,
 
 	t.Logf("Charlie UTXOs after breach: %v", toProtoJSON(t.t, charlieUTXOs))
 }
+
+func testCustomChannelsLiquidityEdgeCases(_ context.Context,
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:1682]: nit: a Godoc with a short description of what the test does would be nice.

File: itest/litd_custom_channels_test.go [hunk 5]
```diff
@@ -1675,3 +1678,207 @@ func testCustomChannelsBreach(_ context.Context, net *NetworkHarness,
...
+
+	logBalance(t.t, nodes, assetID, "after 1k sats")
+
+	// Edge case: The channel reserve check should trigger, and we should
+	// get a payment failure, not a timeout.
+	//
+	// Now Dave tries to send 50 assets to Charlie. There shouldn't be
+	// enough sats in the channel.
+
+	done := make(chan bool, 1)
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:1819]: nit: move to after comment, so there's always a newline before a comment block.

File: itest/litd_custom_channels_test.go [hunk 6]
```diff
@@ -1675,3 +1678,207 @@ func testCustomChannelsBreach(_ context.Context, net *NetworkHarness,
...
+	sendAssetKeySendPayment(
+		t.t, charlie, dave, 50, assetID,
+		fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+	)
+
+	logBalance(t.t, nodes, assetID, "after 50 assets")
+
+	// Normal case.
+	// Send 1k sats from Charlie to Dave.
+	sendKeySendPayment(t.t, charlie, dave, 1000)
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:1809]: Shouldn't there now be 1500 sats on Dave's side? 500 from the asset keysend above (which is hard coded for asset keysends) and the 1k here? And the reserve AFAIK is arount 1060, so there should actually be enough balance for a 354 sat HTLC?

File: itest/litd_custom_channels_test.go [hunk 7]
```diff
@@ -1675,3 +1678,207 @@ func testCustomChannelsBreach(_ context.Context, net *NetworkHarness,
...
+	}()
+
+	select {
+	case <-done:
+	case <-timeoutChan:
+		t.Fatalf("Payment didn't fail within expected time duration")
+
+	}
+
+	logBalance(t.t, nodes, assetID, "after failed 50 assets")
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:1841]: Would be nice if we could also assert the payment's failure reason, that it's actually the `INSUFFICIENT_BALANCE` code. I think we return that somewhere.

**Review by guggero — COMMENTED** [review]
Great to have a specific test for these edge cases :+1:

File: itest/litd_custom_channels_test.go (hunk 6 above)
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:1809]: keysends are using 500 sats, so there's actually not enough! I was following the steps of the bug report [here](https://github.com/lightninglabs/taproot-assets/issues/1060)

File: itest/litd_custom_channels_test.go (hunk 7 above)
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:1841]: yeap good idea, it's already available here through the `SendPayment` RPC response, `result.FailureReason`

Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:1841]: Since this is meant to fail in the `paymentLifecycle` phase, we should get a `NO_ROUTE` instead

The reason the failure happens there is because that's when we actually provide the `500 sats` amount to the `PaymentBandwidth` hook. See more on [this comment](https://github.com/GeorgeTsagk/taproot-assets/blob/tap-chan-liquidity-fixes/tapchannel/aux_traffic_shaper.go#L146-L163)

**Comment by GeorgeTsagk** [comment]
`go.mod` has now been updated to point to the tip of the tapd PR branch which contains the latest [fix](https://github.com/lightninglabs/taproot-assets/pull/1120/commits/335f8eb08b3b7b0a4102edd4f848c684047509d4)

File: itest/litd_custom_channels_test.go (hunk 6 above)
Comment (guggero) [reply itest/litd_custom_channels_test.go:1818]: Ah, right. Might add that as a comment.

**Review by guggero — COMMENTED** [review]
Linter and itest failed, other than that looks good :tada:

**Comment by GeorgeTsagk** [comment]
`go.mod` dependencies updated, after dependent PRs got merged

**Review by guggero — APPROVED** [review]

File: itest/litd_custom_channels_test.go [hunk 8]
```diff
@@ -1675,3 +1684,213 @@ func testCustomChannelsBreach(_ context.Context, net *NetworkHarness,
...
+
+	// Now Dave tries to send 50 assets again, this time he should have
+	// enough sats.
+	sendAssetKeySendPayment(
+		t.t, dave, charlie, 50, assetID,
+		fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+		fn.None[lnrpc.PaymentFailureReason](),
+	)
+
+	logBalance(t.t, nodes, assetID, "after 50 sats backwards")
```
Comment (jharveyb) [inline itest/litd_custom_channels_test.go:1867]: nit: after 50 assets backwards, not sats

**Review by jharveyb — APPROVED** [review]
Great catches, LGTM 🎉
